apps/asap-server/test/handlers/research-output/index-handler.test.ts

```diff
@@ -20,36 +22,379 @@ describe('Research Output index handler', () => {
 
   afterEach(() => jest.clearAllMocks());
 
-  test('Should fetch the research-output and create a record in Algolia when no records exists yet', async () => {
+  test('Should fetch the research-output and create a record in Algolia when research-output is created', async () => {
     const researchOutputResponse = getResearchOutputResponse();
     researchOutputControllerMock.fetchById.mockResolvedValueOnce(
       researchOutputResponse,
     );
 
-    await indexHandler(getEvent());
+    await indexHandler(createEvent('ro-1234'));
 
     expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
       ...researchOutputResponse,
       objectID: researchOutputResponse.id,
     });
   });
+
+  test('Should fetch the research-output and create a record in Algolia when research-output is updated', async () => {
+    const researchOutputResponse = getResearchOutputResponse();
+    researchOutputControllerMock.fetchById.mockResolvedValueOnce(
+      researchOutputResponse,
+    );
+
+    await indexHandler(updateEvent('ro-1234'));
+
+    expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
+      ...researchOutputResponse,
+      objectID: researchOutputResponse.id,
+    });
+  });
+
+  test('Should remove the record Algolia when research-output is unpublished', async () => {
+    const event = unpublishedEvent('ro-1234');
+    await indexHandler(event);
+
+    expect(algoliaIndexMock.deleteObject).toHaveBeenCalledWith(
+      event.detail.payload.id,
+    );
+  });
+
+  test('Should remove the record Algolia when research-output is deleted', async () => {
+    const event = deleteEvent('ro-1234');
+    await indexHandler(event);
+
+    expect(algoliaIndexMock.deleteObject).toHaveBeenCalledWith(
+      event.detail.payload.id,
+    );
+  });
+
+  describe('Should process all events, and not relay on the order', () => {
+    it('receives on correct order the events created and updated', async () => {
+      const roID = 'ro-1234';
+      const createEv = createEvent(roID);
+      const updateEv = updateEvent(roID);
+      const researchOutputResponse = {
+        ...getResearchOutputResponse(),
+        id: roID,
+      };
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
+        ...researchOutputResponse,
+        title: 'Original title',
+      });
+
+      await indexHandler(createEv);
+      expect(algoliaIndexMock.deleteObject).not.toHaveBeenCalled();
+      expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
+        ...researchOutputResponse,
+        title: 'Original title',
+        objectID: researchOutputResponse.id,
+      });
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
```

Comment:
ok so the idea is this - the whole point of this entire block of tests is that we test events that happen close together. What you tested is simply invoking the handler twice when the data changes - you could split your current tests after the first invocation and expectation.

What we want to test is that although we have two different events - the `fetchById` method retyrns the same thing in BOTH invocations. Thats the whole point. We change something in the entity and then we change it again very quickly - so by the time its received by this handler we only see the last version of the entity